backends/graphs/parser.cpp

```diff
@@ -0,0 +1,294 @@
+            auto keyElement = le->components.at(index);
+
+            auto type = typeMap->getType(e, true);
+            int width = type->width_bits();
+            BUG_CHECK(width > 0, "%1%: unknown width", e);
+
+            mpz_class key_value, mask_value;
+            convertSimpleKey(keyElement, key_value, mask_value);
+            unsigned w = 8 * ROUNDUP(width, 8);
+            totalWidth += ROUNDUP(width, 8);
+            value = Util::shift_left(value, w) + key_value;
+            if (mask_value != -1) {
+                mask = Util::shift_left(mask, w) + mask_value;
+                noMask = false;
+            }
+            LOG3("Shifting " << " into key " << key_value << " &&& " << mask_value <<
+                    " result is " << value << " &&& " << mask);
+            index++;
+        }
+
+        if (noMask)
+            mask = -1;
+        return totalWidth;
+    } else {
+        BUG_CHECK(select->components.size() == 1, "%1%: mismatched select/label", select);
+        convertSimpleKey(keySet, value, mask);
+        auto type = typeMap->getType(select->components.at(0), true);
+        return ROUNDUP(type->width_bits(), 8);
+    }
+}
+
+bool ParserGraphs::preorder(const IR::PackageBlock *block) {
+    for (auto it : block->constantValue) {
+        if (it.second->is<IR::ParserBlock>()) {
+            auto name = it.second->to<IR::ParserBlock>()->container->name;
+            pGraph g_;
+            g = &g_;
+            instanceName = boost::none;
+            boost::get_property(g_, boost::graph_name) = name;
+            begin_v = add_vertex("_BEGIN_", VertexType::PARSER);
+            end_v = add_vertex("_END_", VertexType::PARSER);
+            accept_v = add_vertex("accept", VertexType::DEFAULT);
+            parents = {{begin_v, new EdgeUnconditional()}};
+
+            visit(it.second->getNode());
+
+            for (auto parent : defParents)
+                add_edge(parent.first, accept_v, parent.second->label());
+            add_edge(accept_v, end_v, (new EdgeUnconditional())->label());
+            BUG_CHECK(g_.is_root(), "Invalid graph");
+            GraphAttributeSetter()(g_);
+            writeGraphToFile(g_, name);
+        }
+    }
+    return false;
+}
+
+bool ParserGraphs::preorder(const IR::ParserBlock *block) {
+    visit(block->container);
+    return false;
+}
+
+bool ParserGraphs::preorder(const IR::P4Parser *pars) {
+    for (auto state : pars->states) {
+        if (state->name == IR::ParserState::start) {
+            visit(state);
+            break;
+        }
+    }
+    return false;
+}
+
+bool ParserGraphs::preorder(const IR::ParserState *state) {
+    std::stringstream sstream;
+
+    if (state->selectExpression != nullptr) {
+        auto v = add_and_connect_vertex(state->name, VertexType::HEADER);
+        parents.clear();
+
+        if (state->selectExpression->is<IR::PathExpression>()) {
+            parents = {{v, new EdgeSwitch(new IR::DefaultExpression())}};
```

Comment:
these edge classes are not visible from this file; this will break non-unified builds

Reply:
I don't understand what you said? I am pretty new to c++
Can you give little more details, plz. ?

Reply:
Well there is no `EdgeSwitch` declaration visible from this file. It is in controls.cpp. I believe that this is only compiling because we use a unified-build by default, which merges all the source files in this directory before compiling them. Some people build this code in a non-unified way, it which case it would fail. You should try compiling without a unified build (run cmake with `UNIFIED_COMPILATION_ENABLED` set to false). It really should be a regression for this repo IMO but it's not.